src/paymentdisclosure.h

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) 2017 The Zcash developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef ZCASH_PAYMENTDISCLOSURE_H
+#define ZCASH_PAYMENTDISCLOSURE_H
+
+#include "uint256.h"
+#include "clientversion.h"
+#include "serialize.h"
+#include "streams.h"
+#include "version.h"
+
+// For JSOutPoint
+#include "wallet/wallet.h"
+
+#include <cstdint>
+#include <string>
+
+
+#define PAYMENT_DISCLOSURE_VERSION_EXPERIMENTAL 0
+
+typedef JSOutPoint PaymentDisclosureKey;
+
+struct PaymentDisclosureInfo {
+    uint8_t version;          // 0 = experimental, 1 = first production version, etc.
+    uint256 esk;              // zcash/NoteEncryption.cpp
+    uint256 joinSplitPrivKey; // primitives/transaction.h
+    // ed25519 - not tied to implementation e.g. libsodium, see ed25519 rfc
+
+    libzcash::PaymentAddress zaddr;
+
+    PaymentDisclosureInfo() : version(PAYMENT_DISCLOSURE_VERSION_EXPERIMENTAL) {
+    }
+
+    PaymentDisclosureInfo(uint8_t v, uint256 esk, uint256 key, libzcash::PaymentAddress zaddr) : version(v), esk(esk), joinSplitPrivKey(key), zaddr(zaddr) { }
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion) {
+        READWRITE(version);
+        READWRITE(esk);
+        READWRITE(joinSplitPrivKey);
+        READWRITE(zaddr);
+    }
+
+    std::string ToString() const;
+
+    friend bool operator==(const PaymentDisclosureInfo& a, const PaymentDisclosureInfo& b) {
+        return (a.version == b.version && a.esk == b.esk && a.joinSplitPrivKey == b.joinSplitPrivKey && a.zaddr == b.zaddr);
+    }
+
+    friend bool operator!=(const PaymentDisclosureInfo& a, const PaymentDisclosureInfo& b) {
+        return !(a == b);
+    }
+
+};
+
+
+struct PaymentDisclosurePayload {
+    uint8_t version;        // 0 = experimental, 1 = first production version, etc.
+    uint256 esk;            // zcash/NoteEncryption.cpp
+    uint256 txid;           // primitives/transaction.h
+    size_t js;              // Index into CTransaction.vjoinsplit
+    uint8_t n;              // Index into JSDescription fields of length ZC_NUM_JS_OUTPUTS
```

Comment:
Just to see I understand - in our case n is always in {0,1} ?

Reply:
Yes